File: test/bench/test_repeat_until_success.cpp
/*
 * Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
 * Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "bench/Evaluation.hpp"
#include "bench/RepeatUntilSuccess.hpp"

#include "support/Diagnostics.hpp"
#include "support/TestSupport.hpp"

#include "gtest/gtest.h"

#include <numbers>

namespace {

using mqt::bench::Output;
using mqt::bench::RepeatUntilSuccess;
using mqt::bench::RepeatUntilSuccessOptions;

TEST(RepeatUntilSuccess, HasTheOutput) {
  const auto benchmark = ::mqt::test::value(RepeatUntilSuccess::create());
  EXPECT_EQ(benchmark.output(), (Output{"result", 1}));
}

TEST(RepeatUntilSuccess, ValidatesDataWidth) {
  EXPECT_EQ(
      ::mqt::test::value(RepeatUntilSuccess::create()).options().dataQubits,
      1U);
  EXPECT_EQ(::mqt::test::value(RepeatUntilSuccess::create({.dataQubits = 5}))
                .options()
                .dataQubits,
            5U);
  EXPECT_NO_THROW(::mqt::test::value(RepeatUntilSuccess::create(
      {.dataQubits = RepeatUntilSuccessOptions::MAX_DATA_QUBITS})));
  EXPECT_EQ(::mqt::test::errorKind(
                [&] { return RepeatUntilSuccess::create({.dataQubits = 0}); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(
      ::mqt::test::errorKind([&] {
        return RepeatUntilSuccess::create(
            {.dataQubits = RepeatUntilSuccessOptions::MAX_DATA_QUBITS + 1});
      }),
      ::mqt::ErrorCategory::InvalidArgument);
}

TEST(RepeatUntilSuccess, HasThePhaseSensitiveReference) {
  const auto benchmark = ::mqt::test::value(RepeatUntilSuccess::create());
  constexpr auto bias = std::numbers::sqrt2 / 3.;
  EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability("0")), 0.5 + bias);
  EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability("1")), 0.5 - bias);
  EXPECT_EQ(::mqt::test::errorKind([&] { return benchmark.probability(""); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] { return benchmark.probability("00"); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] { return benchmark.probability("x"); }),
            ::mqt::ErrorCategory::InvalidArgument);
}

TEST(RepeatUntilSuccess, EvaluatesTheReferenceWithoutASuccessOutcome) {
  const auto benchmark = ::mqt::test::value(RepeatUntilSuccess::create());
  constexpr auto bias = std::numbers::sqrt2 / 3.;
  const auto allZero = ::mqt::test::value(benchmark.evaluate({{"0", 10}}));
  EXPECT_NEAR(allZero.totalVariationDistance, 0.5 - bias, 1e-15);
  EXPECT_NEAR(allZero.squaredHellingerFidelity, 0.5 + bias, 1e-15);
  EXPECT_FALSE(allZero.successProbability);
}

} // namespace

File: test/python/bench/test_repeat_until_success.py
# Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
# Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Python bindings for the repeat-until-success benchmark."""

from __future__ import annotations

import json
import math

import pytest

from mqt.core.bench import repeat_until_success

from .utils import assert_generates


def test_repeat_until_success_reference() -> None:
    """Expose the output and phase-sensitive reference distribution."""
    benchmark = repeat_until_success.RepeatUntilSuccess()
    assert benchmark.output.name == "result"
    assert benchmark.output.width == 1
    assert benchmark.probability("0") == pytest.approx(0.5 + math.sqrt(2) / 3)
    assert benchmark.probability("1") == pytest.approx(0.5 - math.sqrt(2) / 3)


def test_repeat_until_success_evaluation() -> None:
    """Evaluate deterministic counts against the reference distribution."""
    evaluation = repeat_until_success.RepeatUntilSuccess().evaluate({"0": 10})
    assert evaluation.total_variation_distance == pytest.approx(0.5 - math.sqrt(2) / 3)
    assert evaluation.squared_hellinger_fidelity == pytest.approx(0.5 + math.sqrt(2) / 3)
    assert evaluation.success_probability is None


def test_repeat_until_success_json_roundtrip() -> None:
    """Preserve the benchmark identity through both JSON representations."""
    benchmark = repeat_until_success.RepeatUntilSuccess()
    assert json.loads(benchmark.instance_specification_json)["parameters"] == {}

    instance_copy = repeat_until_success.RepeatUntilSuccess.from_instance_specification_json(
        benchmark.instance_specification_json
    )
    manifest_copy = repeat_until_success.RepeatUntilSuccess.from_manifest_json(benchmark.manifest_json)
    assert instance_copy.case_id == manifest_copy.case_id == benchmark.case_id


def test_repeat_until_success_generation() -> None:
    """Generate a repeat-until-success program."""
    assert_generates(repeat_until_success.RepeatUntilSuccess().generate())
